### PR TITLE
[CPU] enbale GLM_OCR

### PR DESCRIPTION
**Pull request by yuchengliu1**
Enable GLM_OCR on XEON.  

1. add RGB support to AMX Conv3d kernel
2. add mix dtype support to Layernorm
3. support TP=6 in frontend.

The input/weight/bias of Layernorm support fp16/bf16/fp32 and these three dtypes are independent of one another.
output/gate/residual has the same dtype with input.
residual and bias is optional.

Accuracy (using lmms_eval as cookbook advised):
| Tasks  |Filter|n-shot|     Metric      |   |Value|   |Stderr|Stderr_CLT|
|--------|------|-----:|-----------------|---|----:|---|------|---------:|
|ocrbench|none  |     0|ocrbench_accuracy|↑  |0.835|±  |N/A   |    0.0117|

---
### CI States

Latest PR Test (Base): :x: [Run #35314116120](https://github.com/sgl-project/sglang/actions/runs/35314116120)
Latest PR Test (Extra): :x: [Run #35314115941](https://github.com/sgl-project/sglang/actions/runs/35314115941)
Latest PR Test (AMD ROCm 10): :x: [Run #35314115988](https://github.com/sgl-project/sglang/actions/runs/35314115988)

### CODE REVIEW

File: test/registered/cpu/test_conv3d_embed.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+import sgl_kernel  # noqa: F401
+import torch
+import torch.nn.functional as F
+
+from sglang.test.ci.ci_register import register_cpu_ci
+from sglang.test.cpu_test_utils import precision
+from sglang.test.test_utils import CustomTestCase
+
+register_cpu_ci(est_time=10, suite="base-b-test-cpu")
```
Comment (mingfeima) [inline test/registered/cpu/test_conv3d_embed.py:11]:
```suggestion
register_cpu_ci(est_time=10, suite="tage-b-test-cpu-intel")
```

File: python/sglang/kernels/aot/csrc/cpu/norm.cpp [hunk 2]
```diff
@@ -148,39 +148,71 @@ struct NormTraits<NormMode::RMSNormGated> : NormTraitsBase {
...
+    _mm512_storeu_si512(output, (__m512i)(_mm512_cvtne2ps_pbh(value1, value0)));
+  } else {
+    static_assert(std::is_same_v<scalar_t, at::Half>);
+    _mm256_storeu_si256(
+        reinterpret_cast<__m256i*>(output), _mm512_cvtps_ph(value0, _MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC));
+    _mm256_storeu_si256(
+        reinterpret_cast<__m256i*>(output + 16),
+        _mm512_cvtps_ph(value1, _MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC));
+  }
+}
```
Comment (mingfeima) [inline python/sglang/kernels/aot/csrc/cpu/norm.cpp:186]:
```suggestion
```

use `load_float_vec2` + `convert_from_float_ext` from `vec.h`

File: python/sglang/kernels/aot/csrc/cpu/norm.cpp [hunk 3]
```diff
@@ -385,38 +419,35 @@ struct NormReduceGeneric {
...
-  if constexpr (std::is_same_v<scalar_t, at::BFloat16>) {
-    switch (p.D) {
-      LAUNCH_PARALLEL_LOOP_HD(32);
-      LAUNCH_PARALLEL_LOOP_HD(64);
-      LAUNCH_PARALLEL_LOOP_HD(128);
-      LAUNCH_PARALLEL_LOOP_HD(256);
-      LAUNCH_PARALLEL_LOOP_HD(512);
-      default:
-        break;
-    }
```
Comment (mingfeima) [inline python/sglang/kernels/aot/csrc/cpu/norm.cpp:411]: this is a fast path... and you removed it.

**Review by mingfeima — CHANGES_REQUESTED** [review]

**Comment by mingfeima** [comment]
your local base is too old. rebase to master.

File: test/registered/cpu/test_conv3d_embed.py (hunk 1 above)
Comment (yuchengliu1) [reply test/registered/cpu/test_conv3d_embed.py]: I have update it.

File: python/sglang/kernels/aot/csrc/cpu/norm.cpp (hunk 3 above)
Comment (yuchengliu1) [reply python/sglang/kernels/aot/csrc/cpu/norm.cpp:411]: I did not removed it. I removed the line `if` and the indent of `switch` changed.

**Comment by yuchengliu1** [comment]
I have rebase it to master

File: python/sglang/kernels/aot/csrc/cpu/norm.cpp (hunk 2 above)
Comment (yuchengliu1) [reply python/sglang/kernels/aot/csrc/cpu/norm.cpp]: I have change it to load_float_vec2 + convert_from_float_ext

File: python/sglang/kernels/aot/csrc/cpu/norm.cpp [hunk 4]
```diff
@@ -738,18 +750,35 @@ layernorm_cpu(const at::Tensor& input, const at::Tensor& weight, const std::opti
...
   at::Tensor output = at::empty_like(input);
-  AT_DISPATCH_REDUCED_FLOATING_TYPES(st, "layernorm_kernel", [&] {
-    norm4d_kernel_impl<NormMode::LayerNorm, scalar_t>(output.data_ptr<scalar_t>(), input.data_ptr<scalar_t>(), p);
+  AT_DISPATCH_REDUCED_FLOATING_TYPES_AND(at::kFloat, st, "layernorm_input", [&] {
+    using input_t = scalar_t;
+    AT_DISPATCH_REDUCED_FLOATING_TYPES_AND(at::kFloat, weight.scalar_type(), "layernorm_weight", [&] {
+      using weight_t = scalar_t;
+      if (bias.has_value()) {
+        AT_DISPATCH_REDUCED_FLOATING_TYPES_AND(at::kFloat, bias.value().scalar_type(), "layernorm_bias", [&] {
+          using bias_t = scalar_t;
```
Comment (mingfeima) [inline python/sglang/kernels/aot/csrc/cpu/norm.cpp:773]: bias can either be float32 or the same dtype with input.

aka. you will not likely to have input bf16 and bias f16.

what is the real use case in your concerned models?

can you make a table listing :

 - model name
 - input type
 - weight type
 - bias type

Comment (yuchengliu1) [reply python/sglang/kernels/aot/csrc/cpu/norm.cpp:773]: The input/weight/bias of Layernorm support fp16/bf16/fp32 and these three dtypes are independent of one another.
output/gate/residual has the same dtype with input.
residual and bias is optional.

Comment (yuchengliu1) [reply python/sglang/kernels/aot/csrc/cpu/norm.cpp:773]: The models of GLM series have text module and vision module, and they call the different norm OP (They will finally use the kernel modified in this PR).
The type value in table means input_type/weight_type/bias_type.

Model  | text module(RMSnorm) | vision module(layernorm)
-- | -- | --
GLM4.x | BF16/BF16/BF16 | /
GLM4.xV/GLM_ocr | FP16/FP16/FP16 | BF16/FP32/FP32

Comment (mingfeima) [reply python/sglang/kernels/aot/csrc/cpu/norm.cpp:773]: therefore:

-  scalar_t (input) is either bf16 or fp16;
- param_t (weight and bias) is either same as scalar_t or acc_t (fp32)

weight and bias must be the same dtype.

use `CPU_DISPATCH_FLOATING_TYPES_EXT` and `AT_DISPATCH_BOOL`

**Review by mingfeima — CHANGES_REQUESTED** [review]
